src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
 
--- | @since 2.0.0
 instance
-    forall (a :: S -> Type) (b :: S -> Type).
-    ( PLift a
-    , PLift b
+    forall (fin :: S -> Type) (pa :: S -> Type) (pb :: S -> Type).
+    ( PUnLam' fin pb (IsFinal fin pb)
+    , PUnLamHask fin (pa :--> pb) ~ (TestableTerm pa -> PUnLamHask fin pb)
     ) =>
-    FromPFunN (a :--> b) PBool 'LastPFunction
+    PUnLam' fin (pa :--> pb) 'False
     where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) = TestableTerm $ f # x
+    pUnLam' f = \(TestableTerm y) -> pUnLam' @fin @pb @(IsFinal fin pb) (f # y)
 
--- | @since 2.0.0
-instance FromPFunN a PBool 'LastPTerm where
-    fromPFun' _ f (TestableTerm x) = TestableTerm $ f # x
+instance PUnLam' fin fin 'True where
+    pUnLam' x = TestableTerm x
 
--- | @since 2.0.0
-instance
-    forall
-        (a :: S -> Type)
-        (b :: S -> Type)
-        (c :: S -> Type)
-        (d :: S -> Type)
-        (e :: S -> Type).
-    ( c ~ (d :--> e)
-    , PLift a
-    , PLift b
-    , FromPFunN d e (PTT c)
-    ) =>
-    FromPFunN (a :--> b) c 'PFunction
-    where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) =
-        fromPFun' (Proxy @(PTT c)) $ f # x
+{- | Constraint for `PUnLam'` that will give a better type error message.
 
--- | @since 2.0.0
-instance
-    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type) (d :: S -> Type).
-    ( b ~ (c :--> d)
-    , FromPFunN c d (PTT b)
-    ) =>
-    FromPFunN a b 'PTerm
-    where
-    fromPFun' _ f (TestableTerm x) = fromPFun' (Proxy @(PTT b)) $ f # x
+ @since 2.1.0
+-}
+type PUnLam fin p = (PUnLam' fin p (IsFinal fin p), CheckReturn fin p)
+
+{- | Wrapper for @pUnLam'@. Only required the final return type.
+
+ @since 2.1.0
+-}
+punlam' ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam' = pUnLam' @fin @p @(IsFinal fin p)
+
+{- | Wrapper for @pUnLam'@. Same as @punlam'@ but evaluates the given
+     Plutarch function before the conversion. It will throw an error
+     if evaluation failes.
+
+ @since 2.1.0
+-}
+punlam ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam pf = punlam' @fin (loudEval pf)
 
-{- | Converts Plutarch Functions into `Testable` Haskell function of
-  TestableTerms.
+{- | Converts Plutarch Functions into Haskell function of
+     TestableTerms. Then, wrap any instance of TestableTerm that is
+     lambda into `PFun`. Result of this function will be
+     "QuickCheck-compatible" given that all Plutarch type used have
+     `PArbitrary` instance.
 
  @since 2.0.0
 -}
 fromPFun ::
-    forall (a :: S -> Type) (b :: S -> Type) (c :: PTermType).
-    ( c ~ PTT (a :--> b)
-    , FromPFunN a b c
+    forall (p :: S -> Type).
+    ( PUnLam PBool p
+    , PWrapLam (PUnLamHask PBool p)
     ) =>
-    (forall s. Term s (a :--> b)) ->
-    TestableFun c (a :--> b)
-fromPFun = fromPFun' (Proxy @(PTT (a :--> b)))
+    ClosedTerm p ->
```

Comment:
I'm a tad wary of `ClosedTerm` as a function argument: it's a type synonym over a skolemized `s`, which is quite brittle without QuickLook. _With_ QuickLook, we get the issue of breakage of `do`-notation, which is still unresolved.

I'd rather the use of a `newtype` hiding the `s`, but that's _slightly_ out of scope.

Reply:
I really hope `ClosedTerm` itself was a newtype wrapper... But in this case I think this should remain as is because otherwise it will be quite painful to use this function.

Reply:
We'll need to monitor if it gives us really awful errors in practice.